tensorboard/webapp/metrics/views/main_view/card_grid_test.ts

```diff
@@ -0,0 +1,188 @@
+import {CardViewContainer} from '../card_renderer/card_view_container';
+import {CardIdWithMetadata} from '../metrics_view_types';
+import {Store} from '@ngrx/store';
+import {MockStore, provideMockStore} from '@ngrx/store/testing';
+import {appStateFromMetricsState} from '../../testing';
+import {State} from '../../../app_state';
+import {selectors as settingsSelectors} from '../../../settings';
+import * as selectors from '../../../selectors';
+
+const scrollElementHeight = 100;
+
+// Utility funciton to allow use of await.
+function timeout(ms: number) {
+  return new Promise((resolve) => setTimeout(resolve, ms));
+}
+
+@Component({
+  selector: 'testing-component',
+  template: `
+    <div cdkScrollable>
+      <div class="placeholder">placeholder</div>
+      <metrics-card-grid
+        [cardIdsWithMetadata]="cardIdsWithMetadata"
+        [cardObserver]="cardObserver"
+        [groupName]="cardObserver"
+      ></metrics-card-grid>
+      <div class="placeholder">placeholder</div>
+    </div>
+  `,
+  styles: [
+    `
+      div {
+        position: fixed;
+        display: block;
+        height: ${scrollElementHeight}px;
+        overflow-y: scroll;
+      }
+      .placeholder {
+        position: relative;
+        height: 700px;
+        display: block;
+      }
+    `,
+  ],
+})
+class TestableComponent {
+  @Input() groupName: string = 'test group name';
+  @Input() cardIdsWithMetadata: CardIdWithMetadata[] = [
+    {
+      cardId: 'card1',
+      plugin: PluginType.SCALARS,
+      tag: 'tagA',
+      runId: null,
+    },
+    {
+      cardId: 'card2',
+      plugin: PluginType.SCALARS,
+      tag: 'tagA/Images',
+      runId: 'run1',
+      sample: 0,
+    },
+    {
+      cardId: 'card3',
+      plugin: PluginType.SCALARS,
+      tag: 'tagB/meow/cat',
+      runId: 'run1',
+      sample: 0,
+    },
+  ];
+  @Input() cardObserver: CardObserver = new CardObserver();
+}
+
+describe('card grid', () => {
+  let store: MockStore<State>;
+  beforeEach(async () => {
+    await TestBed.configureTestingModule({
+      imports: [NoopAnimationsModule, ScrollingModule],
+      declarations: [
+        CardGridComponent,
+        CardGridContainer,
+        TestableComponent,
+        CardViewComponent,
+        CardViewContainer,
+        CardLazyLoader,
+      ],
+      providers: [
+        provideMockStore({
+          initialState: appStateFromMetricsState(),
+        }),
+      ],
+    }).compileComponents();
+
+    store = TestBed.inject<Store<State>>(Store) as MockStore<State>;
+    store.overrideSelector(selectors.getRunColorMap, {});
+  });
+
+  fit('keeps pagination button position when page size changes', async () => {
+    store.overrideSelector(settingsSelectors.getPageSize, 2);
+    let scrollOffset = Math.floor(Math.random() * scrollElementHeight);
+    const fixture = TestBed.createComponent(TestableComponent);
+    fixture.detectChanges();
+    const nextButtons = fixture.debugElement
```

Comment:
might be more readable to say `const [topNextButton, bottomNextButton] = fixture.debugElement...`

Reply:
Clean. I like it.